### ISSUE TITLE
Fix sample transaction code. was: Difference save entity between using repository and transaction

### ISSUE THREAD

**Issue by rbleuse** [spring-projects/spring-data-couchbase#1524]
Hello,

I was playing with transactions provided by couchbase and spring data following the documentation from [here](https://docs.spring.io/spring-data/couchbase/docs/current/reference/html/#couchbase.transactions)
I noticed that there's a difference if I persist a document with a repository or with a transaction.

Here are details :
spring-boot 2.7.2
couchbase-transactions 1.2.4

Simple entity
```kt
@Document
@Scope("dev")
@Collection("schedule")
data class Schedule(
    @field:Id
    val id: String,

    val nested: Nested,

    @field:Field
    val list: List<String>
)

data class Nested(
    @field:Field("isFree")
    val free: Boolean
)

```

Repository
```kt
interface ScheduleRepository : ReactiveCouchbaseRepository<Schedule, String>
```

Service
```kt
private val collection = clientFactory.withScope("dev").getCollection("schedule").reactive()

fun saveScheduleWithRepo(schedule: Schedule) = repository.save(schedule)

fun saveScheduleAsDocWithTransaction(schedule: Schedule): Mono<TransactionResult> {
    val target = CouchbaseDocument()
    converter.write(schedule, target)
    return transactions.reactive().run {
        it.insert(collection, target.id, target.content)
            .then()
    }
}
```

```kt
val schedule = Schedule("1", Nested(true), listOf("TEST"))
return service.saveScheduleWithRepo(schedule)
    .flatMap {
        service.saveScheduleAsDocWithTransaction(schedule.copy(id = "2"))
    }
```

Here are differences :

<table>
<tr>
<td> with repository</td> <td> as doc with transaction</td>
</tr>
</tr>
<tr>
<td>

```json
{
  "_class": "com.rbleuse.spring.reactive.couchbase.model.Schedule",
  "list": [
    "TEST"
  ],
  "nested": {
    "isFree": true
  }
}
```

</td>
<td>

```json
{
  "_class": "com.rbleuse.spring.reactive.couchbase.model.Schedule",
  "list": {
    "empty": false
  },
  "nested": {
    "content": {
      "isFree": true
    },
    "id": null,
    "expiration": 0
  }
}
```

</td>

</td>
</table>

As you can see, inserting the document with a transaction adds extra fields to the document : `content`, `id` and `expiration` (all in `nested`). Shouldn't it be in root object instead ?
Furthermore, my simple list of string is not converted into an array with the actual String value.
If I convert my List into an Array in my data class, result is same when using a transaction.

Did I miss something ?

**Comment by mikereiche** [comment]
Before I go into investigating the differences -   couchbase-transactions are directly supported in spring-data-couchbase in 5.0.0-M5 [ the doc is staged at https://mikereiche.github.io/staged/ ]  It's not necessary to do the conversion.  There are instructions on how to use a SNAPSHOT at https://docs.spring.io/spring-data/couchbase/docs/5.0.0-M5/reference/html/ (just use milestone instead of snapshot and use M5).

**Comment by mikereiche** [comment]
you need to use target.export() to get the correct serialization of a List.

I made this for a reproducer (I also added a List[] nicknames to User) and it comes out ok.  Your output looks like the last arg to ctx.insert might be target instead of target.content (?).

```
public class User extends AbstractUser implements Serializable {
	@Version protected long version;
	@Transient protected String transientInfo;
	@CreatedBy protected String createdBy;
	@CreatedDate protected long createdDate;
	@LastModifiedBy protected String lastModifiedBy;
	@LastModifiedDate protected long lastModifiedDate;
	List nicknames = List.of("Happy", "Sleepy");
```

```
	public void saveScheduleWithRepo(User user) {
		userRepository.save(user);
	}

	public void saveScheduleAsDocWithTransaction(User user) {
		CouchbaseDocument target = new CouchbaseDocument();
		couchbaseTemplate.getConverter().write(user, target);
		com.couchbase.client.java.Collection col = couchbaseTemplate.getCouchbaseClientFactory().getDefaultCollection();
		couchbaseTemplate.getCouchbaseClientFactory().getCluster().transactions()
				.run((ctx) -> ctx.insert(col, target.getId(), target.export()));
	}

	@Test
	void insert() {
		User u1 = new User(UUID.randomUUID().toString(), "non-trans", "last");
		User u2 = new User(UUID.randomUUID().toString(), "Transaction", "TRANSACTION");
		saveScheduleWithRepo(u1);
		saveScheduleAsDocWithTransaction(u2);
	}
```

```
{
  "_class": "abstractuser",
  "createdBy": "nonreactive_auditor",
  "createdDate": 1577865600000,
  "firstname": "non-trans",
  "lastModifiedBy": "nonreactive_auditor",
  "lastModifiedDate": 1577865600000,
  "lastname": "last",
  "nicknames": [
    "Happy",
    "Sleepy"
  ]
}
```

```
{
  "_class": "abstractuser",
  "createdDate": 0,
  "firstname": "Transaction",
  "lastModifiedDate": 0,
  "lastname": "TRANSACTION",
  "nicknames": [
    "Happy",
    "Sleepy"
  ]
}
```

**Comment by rbleuse** [comment]
Thanks a lot for your suggestion, it's indeed working fine with `target.export()`

I took this code from spring data couchbase documentation, which gives an example using the target.content instead. Perhaps the documentation needs to be updated while waiting for the 5.0.0 to be released

> 8.3. Object Conversions
> Since the transactions library itself has no knowledge of your spring data entity types, you need to convert it back and forth when reading/writing to interact properly. Fortunately, all you need to do is autowire the MappingCouchbaseConverter and utilize it:
> 
> Example 92. Transaction Conversion on Write
> ```java
> @Autowired
> MappingCouchbaseConverter mappingCouchbaseConverter;
> 
> public void doSomething() {
>   transactions.run(ctx -> {
> 
>    Airline airline = new Airline("demo-airline", "at");
>    CouchbaseDocument target = new CouchbaseDocument();
>    mappingCouchbaseConverter.write(airline, target);
> 
>    ctx.insert(couchbaseClientFactory.getDefaultCollection(), target.getId(), target.getContent());
> 
>    ctx.commit();
>   });
> }
> ```

I also tried to use the 5.0.0-M5 and I'm facing this issue :

```kt
@Service
class ScheduleService(
    private val reactiveOperations: ReactiveCouchbaseOperations
) {
    @Transactional
    fun saveScheduleWithTransaction(schedule: Schedule): Mono<Schedule> {
        return reactiveOperations.insertById(Schedule::class.java).one(schedule)
    }
}
```

```
java.lang.IllegalStateException: block()/blockFirst()/blockLast() are blocking, which is not supported in thread reactor-http-nio-3
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:83) ~[reactor-core-3.5.0-M4.jar:3.5.0-M4]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ org.springframework.boot.actuate.metrics.web.reactive.server.MetricsWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ HTTP POST "/schedule" [ExceptionHandlingWebHandler]
Original Stack Trace:
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:83) ~[reactor-core-3.5.0-M4.jar:3.5.0-M4]
		at reactor.core.publisher.Mono.block(Mono.java:1675) ~[reactor-core-3.5.0-M4.jar:3.5.0-M4]
		at org.springframework.data.couchbase.transaction.CouchbaseCallbackTransactionManager.handlePropagation(CouchbaseCallbackTransactionManager.java:191) ~[spring-data-couchbase-5.0.0-M5.jar:5.0.0-M5]
		at org.springframework.data.couchbase.transaction.CouchbaseCallbackTransactionManager.execute(CouchbaseCallbackTransactionManager.java:76) ~[spring-data-couchbase-5.0.0-M5.jar:5.0.0-M5]
		at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:417) ~[spring-tx-6.0.0-M5.jar:6.0.0-M5]
		at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-6.0.0-M5.jar:6.0.0-M5]
		at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.0.0-M5.jar:6.0.0-M5]
		at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-6.0.0-M5.jar:6.0.0-M5]
		at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708) ~[spring-aop-6.0.0-M5.jar:6.0.0-M5]
		at com.rbleuse.spring.reactive.couchbase.service.ScheduleService$$EnhancerBySpringCGLIB$$f091fb23.saveScheduleWithTransaction(<generated>) ~[main/:na]
		at com.rbleuse.spring.reactive.couchbase.handler.ScheduleHandler.createSchedule(ScheduleHandler.kt:20) ~[main/:na]
```

**Comment by mikereiche** [comment]
the blocking issue is fixed in https://github.com/spring-projects/spring-data-couchbase/issues/1516
It should already be in the snapshot.  

https://repo.spring.io/ui/native/libs-snapshot/org/springframework/data/spring-data-couchbase/5.0.0-SNAPSHOT/

**Comment by mikereiche** [comment]
> I also tried to use the 5.0.0-M5 and I'm facing this exception (it's same with the snapshot one)

I'll see if I can eliminate that block().

I'm curious how this code is being called as you had a similar issue with #1516 and I haven't been able to reproduce it. (I did provide a fix for #1516 which eliminates the block()).
